src/matcher.rs

```diff
@@ -0,0 +1,160 @@
+/// Check if input is a valid domain label
+pub fn is_label(input: &str) -> bool {
+    if input.is_empty() {
```

Comment:
DNS labels also have maximum length of 63 octets ([RFC1035 2.3.4](https://tools.ietf.org/html/rfc1035#section-2.3.4)) so we could check that here also?

```suggestion
    if input.is_empty() || input.len > 63 {
```

Reply:
Sure, I'll add!

Reply:
Thinking about this in terms of code separation. The number of DNS labels are checked here at the call site: https://github.com/rushmorem/publicsuffix/blob/master/src/lib.rs#L613

The empty check is required for the following logic to be correct, but I'm wondering whether the 63 octet check should be done where we call `matcher::is_label` instead?

Reply:
Turns out this is already checked for here in `verify_dns_length`:

https://github.com/rushmorem/publicsuffix/blob/master/src/lib.rs#L706

```rust
    fn to_ascii(domain: &str) -> Result<String> {
        let result = idna::Config::default()
            .transitional_processing(true)
            .verify_dns_length(true)
            .to_ascii(domain);
        result.map_err(|error| ErrorKind::Uts46(error).into())
    }
```

And there's a test checking for it here: https://github.com/rushmorem/publicsuffix/blob/master/src/tests.rs#L148

It might be better relying on this mechanism, cause I'm a bit uncertain about ASCII vs UTF-8 here. The idna package should do the right thing with regards to that though.

Reply:
DNS addresses (aka hostnames) have two limits: The total length and the label length. The amount of labels can then be derived from this: 253 chars / 2 = 126.5 ~= 127. 

`is_label()` is validating a DNS label and the specification says the maximum length is 63. So for me it feels it's the proper place to do the validation of this maximum length.

And you are right about `idna` crate. It does the validation, but I found it to be slow when given anomalous input (from fuzzing). Issue here: https://github.com/servo/rust-url/issues/595

And IDNA (ASCII vs. Unicode hostnames) makes things more difficult. I'm trying to figure out best way to do the most cheap and simple checks first before going to expensive conversions (like IDNS to_ascii()) to validate everything.

Reply:
Alright, I don't mind hugely where the test happens, however as long as that slow `idna` test is there, maybe we shouldn't double up the logic?

Maybe cleaner to make a commit that both adds the new check – with proper thinking around ASCII vs Unicode – and disable the `idna` at the same time?